malcolm/modules/scanning/controllers/runnablecontroller.py

```diff
@@ -466,7 +470,15 @@ def do_run(self, hook):
         self.abortable_transition(ss.POSTRUN)
         completed_steps = self.configured_steps.value
         if completed_steps < self.total_steps.value:
-            steps_to_do = self.steps_per_run
+            if len(self.steps_per_run) > 1:
+                if self._breakpoint < len(self.steps_per_run):
+                    steps_to_do = self.steps_per_run[self._breakpoint]
+                    self._breakpoint += 1
+                # handle a case when the endpoint is not provided
+                else:
```

Comment:
What happens in this case if we aren't being asked to move all axes?

Reply:
This is now refactored. Now if the last breakpoint doesn't finish the scan an additional breakpoint is added to complete the inner scan. This is done before multiplying the list of breakpoints for each outer step.